forc-pkg/src/source/mod.rs

```diff
@@ -145,37 +145,65 @@ pub struct DisplayCompiling<'a, T> {
 pub struct PinnedParseError;
 
 impl Source {
+    /// Construct a source from path information collected from manifest file.
+    fn with_path_dependency(
+        relative_path: &Path,
+        manifest_dir: &Path,
+        member_manifests: &MemberManifestFiles,
+    ) -> Result<Self> {
+        let path = manifest_dir.join(relative_path);
+        let canonical_path = path
+            .canonicalize()
+            .map_err(|e| anyhow!("Failed to canonicalize dependency path {:?}: {}", path, e))?;
+        // Check if path is a member of a workspace.
+        if member_manifests
+            .values()
+            .any(|pkg_manifest| pkg_manifest.dir() == canonical_path)
+        {
+            Ok(Source::Member(member::Source(canonical_path)))
+        } else {
+            Ok(Source::Path(canonical_path))
+        }
+    }
+
+    /// Construct a source from version information collected from manifest file.
+    fn with_version_dependency(
+        pkg_name: &str,
+        version: &str,
+        namespace: &reg::file_location::Namespace,
+    ) -> Result<Self> {
+        // TODO: semver parsing does not work for versions like: 0.1 or ^0.1.
```

Comment:
interesting; i wonder if we can ensure upstream that versions should always be semver compliant

Reply:
At first i thought there was no out of box option for this but in a recent discussion, we found out that there is a [VersionReq](https://docs.rs/semver/1.0.26/semver/struct.VersionReq.html) I will switch to that in the follow up PR which will also add the logic for handling the caret and wildcard type of declarations. So i am keeping the parsing step like this as the logic to select correct version is still going to come in the next PR.

Reply:
Can we leave a link to the github issue for this here?

Reply:
Done ine [a834871](https://github.com/FuelLabs/sway/pull/7038/commits/a834871f5a097a293ea5bc053229d0ab8eccf566)